### PR TITLE
e2ee: commit a store transaction just after generating one-time keys

### PR DESCRIPTION
**Pull request by bnjbvr**
See commit messages for details. I've gotten several EX devices wedged, and this is my biggest guess so far why it has happened in the first place.

### CODE REVIEW

**Review by jplatte — APPROVED** [review]
Apart from the fact that I've never heard the term "applicative transaction", looks good.

**Comment by bnjbvr** [comment]
> Apart from the fact that I've never heard the term "applicative transaction", looks good.

You're right, this is made up vocabulary as of today, to disambiguate from sqlite- (or indexeddb, for that matter) level transactions.

**Comment by jplatte** [comment]
Maybe "application-level"? "applicative" has an existing meaning [in functional programming](https://hackage.haskell.org/package/base-4.19.0.0/docs/Control-Applicative.html).

**Comment by bnjbvr** [comment]
Ah yeah, I would argue people would think more immediately of applicative as "relating to an application" :-)

**Review by uhoreg — APPROVED** [review]
Flushing a transaction seems to be contrary to how transactions are normally supposed to work (all-or-nothing), but I guess it works here.

File: crates/matrix-sdk-crypto/src/machine.rs [hunk 1]
```diff
@@ -1211,7 +1211,20 @@ impl OlmMachine {
             account.update_key_counts(
                 sync_changes.one_time_keys_counts,
                 sync_changes.unused_fallback_keys,
-            )
+            );
+
+            // Flush the applicative transaction as soon as the keys have been generated. If
+            // we don't do it here, then it's possible that the
+            // `StoreTransaction` fails, but we've uploaded keys in the
```
Comment (BillCarsonFr) [inline crates/matrix-sdk-crypto/src/machine.rs:1218]: How would it happen? If the store transaction fails, then the `receive_sync_changes` will fail, then the whole sync will fail and the keys wont be uploaded?

I think that in that case the sync token wouldn't be persisted, so the app will try again the same sync?

**Review by BillCarsonFr — CHANGES_REQUESTED** [review]

**Review by Hywan — APPROVED** [review]
Looks good to me, thanks!

**Comment by bnjbvr** [comment]
Discussed on chat, it's probably not a fix to the issue I've observed, since the OTKs aren't uploaded at the same time a store transaction is alive. Instead, we'll add more logs to try to understand which keys are generated twice.

**Comment by richvdh** [comment]
This was an attempt to fix https://github.com/matrix-org/matrix-rust-sdk/issues/1415

**Pull request closed without merging**